crates/environ/src/fact/trampoline.rs

```diff
@@ -80,50 +85,168 @@ struct Compiler<'a, 'b> {
 }
 
 pub(super) fn compile(module: &mut Module<'_>, adapter: &AdapterData) {
+    fn compiler<'a, 'b>(
+        module: &'b mut Module<'a>,
+        adapter: &AdapterData,
+    ) -> (Compiler<'a, 'b>, Signature, Signature) {
+        let lower_sig = module.types.signature(&adapter.lower, Context::Lower);
+        let lift_sig = module.types.signature(&adapter.lift, Context::Lift);
+        let ty = module
+            .core_types
+            .function(&lower_sig.params, &lower_sig.results);
+        let result = module
+            .funcs
+            .push(Function::new(Some(adapter.name.clone()), ty));
+
+        // If this type signature contains any borrowed resources then invocations
+        // of enter/exit call for resource-related metadata tracking must be used.
+        // It shouldn't matter whether the lower/lift signature is used here as both
+        // should return the same answer.
+        let emit_resource_call = module.types.contains_borrow_resource(&adapter.lower);
+        assert_eq!(
+            emit_resource_call,
+            module.types.contains_borrow_resource(&adapter.lift)
+        );
+
+        (
+            Compiler {
+                types: module.types,
+                module,
+                code: Vec::new(),
+                nlocals: lower_sig.params.len() as u32,
+                free_locals: HashMap::new(),
+                traps: Vec::new(),
+                result,
+                fuel: INITIAL_FUEL,
+                emit_resource_call,
+            },
+            lower_sig,
+            lift_sig,
+        )
+    }
+
+    let start_adapter = |module: &mut Module, param_globals| {
```

Comment:
Mind throwing "async" in the name of these local variables (and `return_adapter` below) to clarify they're async-related things?